### PR TITLE
fix!: migrate to EventEmitterAsyncResource

### LINKED ISSUE #82: Installing tinypool causes typescript errors.
Tinypool is conflicting so some dependency:
```
Error: node_modules/tinypool/dist/index.d.ts(21,15): error TS2417: Class static side 'typeof EventEmitterAsyncResource' incorrectly extends base class static side 'typeof EventEmitter'.
  Types of property 'EventEmitterAsyncResource' are incompatible.
    Types of construct signatures are incompatible.
      Type 'new (options?: string | Options$1 | undefined) => EventEmitterAsyncResource' is not assignable to type 'new (options?: EventEmitterAsyncResourceOptions | undefined) => EventEmitterAsyncResource'.
        Construct signature return types 'EventEmitterAsyncResource' and 'EventEmitterAsyncResource' are incompatible.
          The types of 'asyncId' are incompatible between these types.
            Type '() => number' is not assignable to type 'number'.
Error: node_modules/tinypool/dist/index.d.ts(146,31): error TS2304: Cannot find name 'Worker'.
```

CI run: https://github.com/mshima/generator-jhipster/actions/runs/7835634934/job/21381450038?pr=914

### PR DESCRIPTION
**Pull request by userquin**
This PR also includes:
- update pnpm to 9.0.6
- update engines to node 18 or >= 20
- update tsup to 8: now all assets in dist and entry folders (removed esm subfolders)
- add `exports`  ~~including `entry/utils` subpackage~~ (only entry point is public)
- removed unused dependencies

closes #82

### CODE REVIEW

File: src/index.ts [hunk 1]
```diff
@@ -3,8 +3,8 @@ import {
   MessagePort,
   receiveMessageOnPort,
 } from 'worker_threads'
-import { once } from 'events'
-import EventEmitterAsyncResource from './EventEmitterAsyncResource'
+import { once, EventEmitterAsyncResource } from 'events'
+// import EventEmitterAsyncResource from './EventEmitterAsyncResource'
```
Comment (userquin) [inline src/index.ts:7]: remove this

File: package.json [hunk 2]
```diff
@@ -1,61 +1,80 @@
...
+  ],
+  "exports": {
+    ".": {
+      "types": "./dist/index.d.ts",
+      "default": "./dist/index.js"
+    },
+    "./entry/utils": {
+      "types": "./dist/entry/utils.d.ts",
+      "default": "./dist/entry/utils.js"
+    }
```
Comment (AriPerkkio) [inline package.json:29]: I think these could be removed here. The only public API we provide is the entrypoint `dist/index.js`.

File: package.json [hunk 3]
```diff
@@ -1,61 +1,80 @@
...
     "prettier": "^2.5.1",
     "regenerator-runtime": "^0.13.9",
     "snazzy": "^9.0.0",
-    "tsup": "^5.11.6",
-    "typescript": "4.3.x",
+    "tsup": "^8.0.2",
+    "typescript": "^5.4.5",
     "vite": "^5.2.10",
     "vitest": "^1.5.0"
   },
```
Comment (AriPerkkio) [inline package.json:77]: Looks like there are plenty of unused dependencies. We should be able to remove all of these as well:

- `@types/concat-stream`
- `concat-stream`
- `@typescript-eslint/eslint-plugin` (ESLint is not set for this repo)
- `@typescript-eslint/eslint-parser`
- `abort-controller`
- `esbuild`
- `esbuild-register`
- `gen-esm-wrapper`
- `regenerator-runtime`
- `snazzy`

File: test/fixtures/notify-then-sleep.ts [hunk 4]
```diff
@@ -1,3 +1,4 @@
+// @ts-ignore
 export default function (i32array) {
```
Comment (AriPerkkio) [inline test/fixtures/notify-then-sleep.ts:2]:
```suggestion
export default function (i32array: Int32Array) {
```

Comment (AriPerkkio) [reply test/fixtures/notify-then-sleep.ts:2]: Or wait, is this file used by any test?

**Review by AriPerkkio — CHANGES_REQUESTED** [review]
Overall looks good. Thanks!

File: test/fixtures/notify-then-sleep.ts (hunk 4 above)
Comment (userquin) [reply test/fixtures/notify-then-sleep.ts:2]: I have no idea, I just review all imports and found some ts errors.

**Comment by userquin** [comment]
There are some ts errors in the tests: `task-queue.test.ts` and `uncaught-exception-from-handler.test.ts`

**Comment by AriPerkkio** [comment]
> There are some ts errors in the tests: `task-queue.test.ts` and `uncaught-exception-from-handler.test.ts`

Typescript is a bit more strict now. I can fix these after merging this PR as it seems I'm unable to push to your fork. I might be missing some permissions from Tinylibs organization. 🤔

**Review by AriPerkkio — APPROVED** [review]
Thanks for the fix and other maintenance related changes! 🙌 

I'll be merging and releasing this in a new breaking change release next week.

**Pull request merged**